### PR TITLE
Climber

### PR DESCRIPTION
**Pull request by CoYoav**

### CODE REVIEW

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -104,6 +104,28 @@ public static final class ModuleConstants {
...
+
+        public static final double kGearing = 0;
+        public static final double kDriveWheelRaius = 0;
+
+        public static final double kRotToMeter = 2 * PI * kDriveWheelRaius * kGearing;
+        public static final double kRotToMeterPerSec = kRotToMeter / 60;
+        public static final int leftID = 0;
+        public static final int rightID = 0;
+        public static final double LEFT_ARM_LOCATION = 0;
+        public static final double RIGHT_ARM_LOCATION = 0;
```
Comment (TapChap) [inline src/main/java/frc/robot/Constants.java]: what are these values? what do they represent? if it's the encoder values then they shouldn't be final or in the constants..

File: src/main/java/frc/robot/Robot.java [hunk 2]
```diff
@@ -67,6 +70,9 @@ private void configureBindings() {
     controller.square().whileTrue(swerve.pathFindToLocation(HM_LEFT));
     controller.triangle().whileTrue(swerve.pathFindToLocation(HM_CENTER));
     controller.circle().whileTrue(swerve.pathFindToLocation(HM_RIGHT));
+
+    //Clim bear
+    Climber.climber.setDefaultCommand(Climber.climber.manualCommand(()-> operator.getLeftY(), ()-> operator.getRightY()));
```
Comment (TapChap) [inline src/main/java/frc/robot/Robot.java]: NO SINGELTONS

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+import frc.lib.Gains;
+import frc.lib.Neo;
+
+import java.util.function.Supplier;
+
+import static frc.robot.Constants.ClimberConstants.*;
+
+public class Climber extends SubsystemBase {
+    ClimberSide leftSide;
+    ClimberSide rightSide;
```
Comment (TapChap) [inline src/main/java/frc/robot/subsystems/Climber.java]: private final? private final..

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    public Command setClimbHeightsCommand(double leftLength, double rightLength) {
+        return new ParallelCommandGroup(
+                leftSide.setLengthCommand(leftLength),
+                rightSide.setLengthCommand(rightLength)
+        );
+    }
+
+    public Command smartOpenCommand(Supplier<Pose3d> robotPoseSupplier){
```
Comment (TapChap) [inline src/main/java/frc/robot/subsystems/Climber.java:31]: do you really need a pose3d??? isnt the pose2d of the robot enough?

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+    public Command smartCloseCommand(Supplier<Pose3d> robotPoseSupplier){
+        return new RunCommand(() -> setLiftForces(getChainRoll(robotPoseSupplier.get())))
+                .alongWith(
+                        setClimbHeightsCommand(
+                                leftSide.getHeight() - getDesiredHeight(),
+                                rightSide.getHeight() - getDesiredHeight()));
+    }
+
+    private double getChainRoll(Pose3d pose) {
+        return 0;//TODO: calculate the robot roll relative to the chain
```
Comment (TapChap) [inline src/main/java/frc/robot/subsystems/Climber.java]: isnt this info you get for the gyro?

**Review by TapChap — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Constants.java [hunk 6]
```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
...
+        //gains
+        public static final Gains LEFT_GAINS = new Gains(0, 0, 0);
+        public static final Gains RIGHT_GAINS = new Gains(0, 0, 0);
+
+        //movement limitation
+        public static final double kMaxLinearVelocity = 0;
+        public static final double kMaxLinearAcceleration = 0;
+
+        public static final double kGearing = 0;
+        public static final double kDriveWheelRaius = 0;
```
Comment (Starlight220) [inline src/main/java/frc/robot/Constants.java]: The climber can drive? Is this a descriptive name?

File: src/main/java/frc/robot/Constants.java [hunk 7]
```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
...
+                this.centerStage = centerStage;
+            }
+
+            public Translation2d getNegEdge() {
+                return negEdge;
+            }
+
+            public Translation2d getPosEdge() {
+                return posEdge;
+            }
```
Comment (Starlight220) [inline src/main/java/frc/robot/Constants.java]: Immutable enum fields should be `public final` rather than having getters.

File: src/main/java/frc/robot/Constants.java [hunk 8]
```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
...
+
+            private boolean inRange(Translation2d robotTranslation) {
```
Comment (Starlight220) [inline src/main/java/frc/robot/Constants.java]: Document both what the function does, and explain the math inside.

File: src/main/java/frc/robot/Constants.java [hunk 9]
```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
...
+                        Math.atan(robotTranslation.minus(centerStage).getX() / robotTranslation.minus(centerStage).getY());
+                double negAngle = Math.atan(negEdge.minus(centerStage).getX() / negEdge.minus(centerStage).getY());
+                double posAngle = Math.atan(posEdge.minus(centerStage).getX() / posEdge.minus(centerStage).getY());
+
+                if (posAngle > negAngle) return robotStageAngle < posAngle && robotStageAngle > negAngle;
+                return (robotStageAngle > posAngle && robotStageAngle > negAngle)
+                        || (robotStageAngle < posAngle && robotStageAngle < negAngle);
+            }
+
+            public static Chain getBestChain(Translation2d robotTranslation) {
```
Comment (Starlight220) [inline src/main/java/frc/robot/Constants.java:168]: 1. Documentation documentation documentation.
2. Early returns might be preferable here than a result variable, but that might be personal preference/style. I prefer it because I find it easier to track when nothing can change the result after it's set.
3. What happens if this function is called from the other side of the field? Is that what we want to happen?

File: src/main/java/frc/robot/Constants.java [hunk 10]
```diff
@@ -104,6 +104,99 @@ public static final class ModuleConstants {
...
+                    else bestChain = Chain.CHAIN_240;
+                } else {
+                    if (Chain.CHAIN_60.inRange(robotTranslation)) bestChain = Chain.CHAIN_60;
+                    else if (Chain.CHAIN_180.inRange(robotTranslation)) bestChain = Chain.CHAIN_180;
+                    else bestChain = Chain.CHAIN_300;
+                }
+                return bestChain;
+            }
+
+            public static Translation2d getProjection(Translation2d robotPose, Chain chain) {
```
Comment (Starlight220) [inline src/main/java/frc/robot/Constants.java]: What does this function do? Documentation documentation documentation.
Also, I'm not entirely sure that a `T.f(..., t)` is better than `t.f(...)` -- it very much depends on semantics that are very unclear here.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+import java.util.function.DoubleSupplier;
+import java.util.function.Supplier;
+
+import static frc.robot.Constants.ClimberConstants.*;
+import static frc.robot.util.AllianceUtils.isBlueAlliance;
+
+public class Climber extends SubsystemBase {
+    ClimberSide leftSide;
+    ClimberSide rightSide;
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: Don't leave fields package-private. Like @TapChap reviewed earlier, these should be private final.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+
+    public Climber() {
+        leftSide = new ClimberSide(LEFT_GAINS, leftID);
+        leftSide = new ClimberSide(RIGHT_GAINS, rightID);
+    }
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: If indeed this is all the initialization there is, this can be inlined into the field definitions.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+
+
+
+    private double getDesiredHeight() {
+        return Math.min(leftSide.getHeight(), rightSide.getHeight()) - MINIMAL_HEIGHT;
+    }
+
+    private class ClimberSide {
+        private Neo motor;
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: `final`
In general. If something _can_ be final, it should.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+        private TrapezoidProfile trapezoidProfile;
+        private Timer timer;
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: These should not be at this scope.

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+                        timer.restart();
+                        trapezoidProfile = new TrapezoidProfile(new TrapezoidProfile.Constraints(kMaxLinearVelocity, kMaxLinearAcceleration));
+                    },
+                    () -> {
+                        TrapezoidProfile.State currentState = new TrapezoidProfile.State(motor.getPosition(), motor.getVelocity());
+                        TrapezoidProfile.State goalState = new TrapezoidProfile.State(length, 0);
+
+                        TrapezoidProfile.State setpoint = trapezoidProfile.calculate(timer.get(), currentState, goalState);
+                        double feedForward = Math.signum(setpoint.velocity) * motor.getGains().ks +
+                                motor.getGains().kv * currentState.velocity + motor.getGains().kg + liftingForce;
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: Any reason of using feed forward manually rather than ElevatorFeedforward?

File: src/main/java/frc/robot/subsystems/Climber.java [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+                        motor.setReference(setpoint.position, CANSparkBase.ControlType.kPosition, 0, feedForward);
+                    },
+                    (__) -> timer.stop(),
+                    () -> timer.hasElapsed(trapezoidProfile.totalTime())
+            );
+        }
+
+        private void setLiftingForce(double liftingForce) {
+            this.liftingForce = liftingForce;
+        }
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber.java]: What is the purpose of this (the setter as well as the field itself) ?
Documentation.

File: src/main/java/frc/robot/subsystems/Climber/Climber.java [hunk 17]
```diff
@@ -0,0 +1,66 @@
+package frc.robot.subsystems.Climber;
```
Comment (Starlight220) [inline src/main/java/frc/robot/subsystems/Climber/Climber.java]: Packages should be lowercase.
Also, why is this class duplicated?
I'm not looking at this separately yet, figure out the duplication and address my comments on the other file.

File: src/main/java/frc/lib/Neo.java [hunk 18]
```diff
@@ -1,45 +1,63 @@
 package frc.lib;
```
Comment (Starlight220) [inline src/main/java/frc/lib/Neo.java:1]: These changes are out of scope.
Up to @TapChap

**Review by Starlight220 — CHANGES_REQUESTED** [review]
In general there seems to be a lot of removing the end-of-file newline for some reason?
Also, this needs to be rebased and the conflicts dealt with.

File: src/main/java/frc/robot/subsystems/Climber.java (hunk 16 above)
Comment (CoYoav) [reply src/main/java/frc/robot/subsystems/Climber.java]: The purpose is to adjust to the forces applied to the climber when climbing, and the robot's weight is carried , I already have the physics to calculate it using the gyro, I only need to update it

File: src/main/java/frc/robot/Constants.java [hunk 19]
```diff
@@ -104,6 +104,104 @@ public static final class ModuleConstants {
...
+        public static final double kMaxLinearAcceleration = 0;
+
+        public static final double kGearing = 0;
+        public static final double kWinchRadius = 0;
+
+        public static final double kRotToMeter = 2 * PI * kWinchRadius * kGearing;
+        public static final double kRotToMeterPerSec = kRotToMeter / 60;
+        public static final int leftID = 0;
+        public static final int rightID = 0;
+        public static final double kG = 0; //the weight of the robot
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/Constants.java]: please use the proper Constant naming convention

File: src/main/java/frc/robot/Constants.java [hunk 20]
```diff
@@ -104,6 +104,104 @@ public static final class ModuleConstants {
...
+                if (isBlueAlliance()) {
+                    if (Chain.CHAIN_0.inRange(robotTranslation)) return Chain.CHAIN_0;
+                    else if (Chain.CHAIN_120.inRange(robotTranslation)) return Chain.CHAIN_120;
+                    return Chain.CHAIN_240;
+                }
+
+                if (Chain.CHAIN_60.inRange(robotTranslation)) return Chain.CHAIN_60;
+                else if (Chain.CHAIN_180.inRange(robotTranslation)) return Chain.CHAIN_180;
+                return Chain.CHAIN_300;
+            }
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/Constants.java:179]: this wouldn't be for the robot to decide on it's own, the driver will choose the chain to climb on depending on other robots, and other factors,

File: src/main/java/frc/robot/Robot.java [hunk 21]
```diff
@@ -67,6 +68,9 @@ private void configureBindings() {
     controller.square().whileTrue(swerve.pathFindToLocation(HM_LEFT));
     controller.triangle().whileTrue(swerve.pathFindToLocation(HM_CENTER));
     controller.circle().whileTrue(swerve.pathFindToLocation(HM_RIGHT));
+
+    //Clim bear
+    //Climber.climber.setDefaultCommand(Climber.climber.manualCommand(()-> operator.getLeftY(), ()-> operator.getRightY()));
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/Robot.java]: you are not creating an instance of the climber

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 22]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    public Command smartCloseCommand(DoubleSupplier rollSupplier) {
+        //returns a command that lifts the robot of the ground
+        //while making sure it is balanced with regard of the forces applied to each arm separately
+        return new RunCommand(() -> setLiftForces(rollSupplier.getAsDouble()))
+                .alongWith(
+                        setClimbHeightsCommand(
+                                leftSide.getHeight() - getDesiredHeight(),
+                                rightSide.getHeight() - getDesiredHeight()));
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: please refactor this a little and make it look atleast half decend

File: src/main/java/frc/robot/subsystems/climber/Climber.java [hunk 23]
```diff
@@ -0,0 +1,87 @@
...
+                            kG * (TRACK_WIDTH / 2 * Math.cos(-chainRollRad) - leftSide.getHeight() * Math.sin(-chainRollRad)) /
+                                    ((rightSide.getHeight() + leftSide.getHeight()) * Math.sin(-chainRollRad))
+            );
+
+            rightSide.setLiftingForce(
+                    -kG * (TRACK_WIDTH / 2 * Math.cos(-chainRollRad) - leftSide.getHeight() * Math.sin(-chainRollRad)) /
+                            ((rightSide.getHeight() + leftSide.getHeight()) * Math.sin(-chainRollRad))
+            );
+        }
+    }
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/subsystems/climber/Climber.java]: bruhhh, that's soo messy

File: src/main/java/frc/robot/subsystems/climber/ClimberSide.java [hunk 24]
```diff
@@ -0,0 +1,58 @@
...
+                //a check if the time to open the arm ended
+                () -> timer.hasElapsed(trapezoidProfile.totalTime())
+        );
+    }
+
+
+    public void setLiftingForce(double liftingForce) {
+        //sets the T force required to lift the robot at the time in this arm
+        this.liftingForce = liftingForce;
+    }
```
Comment (ExcaliburFRC6738) [inline src/main/java/frc/robot/subsystems/climber/ClimberSide.java:54]: empty line

**Review by ExcaliburFRC6738 — CHANGES_REQUESTED** [review]
I reviewed mostly the code itself, and not really the physical and mathematical calculations, I'm leaving it to @Starlight220 (if you want).

**Comment by ExcaliburFRC6738** [comment]
was merged to a different branch
